fgpyo/io/__init__.py

```diff
@@ -0,0 +1,198 @@
+
+    Args:
+        path: a Path to check
+
+    Example:
+    >>> assert_file_exists(path = Path("some_file.csv"))
+    """
+    assert path.exists(), f"Cannot read non-existent path: {path}"
+    assert not path.is_dir(), f"Cannot read path becasue it is a directory: {path}"
+    assert os.access(path, os.R_OK), f"Path exists but is not readable: {path}"
+
+
+def assert_directory_exists(path: Path) -> None:
+    """Asserts that a path exist and is a directory
+
+    Args:
+        path: Path to check
+
+    Example:
+    >>> assert_directory_exists(path = Path("/example/directory/"))
+    """
+    assert path.exists(), f"Path does not exist: {path}"
+    assert path.is_dir(), f"Path exists but is not a directory: {path}"
+
+
+def assert_path_is_writeable(path: Path, parent_must_exist: bool = True) -> Any:
+    """Asserts the following:
+    If the file exists then it must also be writeable. Else if the path is not a file and
+    parent_must_exist is true then assert that the parent directory exists and is writeable.
+    Else if the path is not a directory and parent_must_exist is false then look at each parent
+    directory until one is found that exists and is writeable.
+
+    In each scenario if a writeable file/directory is found the path is returned if not an
+    AssertionError is raised.
+
+    Args:
+         path: Path to check
+         parent_must_exist: True/False the parent directory must exist, the default is True
+
+    Example:
+    >>> assert_path_is_writeable(path = Path("example.txt"))
+    // Assuming example.txt exists and is writeable
+    >>> example.txt
+    """
+    # If file exists, it must be writeable
+    if path.is_file():
+        assert os.access(path, os.W_OK), f"File exists but is not writebale: {path}"
+        return path
+
+    # Else if file doesnt exist and parent_must_exist is True then check
+    # that path.absolute().parent exists, is a directory and is writeable
+    elif path.is_file() is False & parent_must_exist is True:
+        assert (
+            path.absolute().parent.exists()
+            & path.absolute().parent.is_dir()
+            & os.access(path.absolute().parent, os.W_OK)
+        ), f"File does not have a writeable parent directory: {path}"
+        return path.absolute().parent
+
+    # Else if file doesn't exist and parent_must_exist is False, test parent until
+    # you find the first extent path, and check that it is a directory and is writeable.
+    elif path.is_dir() & parent_must_exist is False:
+        for directory in list(path.parents):
+            try:
+                assert (
+                    directory.is_dir() & directory.exists() & os.access(directory, os.W_OK)
+                ), f"File does not have a writeable parent directory: {path}"
+                return directory
+            except AssertionError:
+                pass
+        raise AssertionError(f"No parent directories of the path provided are writeable: {path}")
```

Comment:
This is super confusing, and I still don't think it does the right thing.  What happens if you have `/A/B/C/D.txt` as your `path`, and `C` exists and is non-writable, but `B` also exists and is writable?

Instead this should be something more like:

```python
for p in path.parents:
    if p.exists():
        assert p.is_dir() and os.access(directory, os.W_OK), f"..."
        return
raise AssertionError(f"No parent directories exist for: {path}")
```